llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

```diff
@@ -1246,6 +1246,14 @@ void DwarfDebug::finishSubprogramDefinitions() {
   }
 }
 
+void DwarfDebug::finalizeAccelerationTables() {
+  for (auto &Entry : AccelDebugNames.getEntries()) {
+    for (AccelTableData *Value : Entry.second.Values) {
+      static_cast<DWARF5AccelTableData *>(Value)->normalizeDIEToOffset();
+    }
+  }
+}
```

Comment:
Yeah, mixed feelings - I wouldn't mind this being a member function of `AccelTable` - and having some function on all entry data types - which can be a no-op on AppleNames entry types (though that'll be a bit subtle - it's only a valid no-op because the CU DIEs do persist long enough to be usable - but AppleNames will hopefully eventually go away, so that's got a path to removal, hopefully).

I did also realize that in the `DWARF5AccelTableStaticData` looks a lot like the new version of the `DWARF5AccelTableData` you're introducing here - so there might be a chance to merge these, especially once the packing/structure layout/size issue is addressed (by not using `std::variant` the struct could be shrunk down to the same size/basically the same layout (making the tag 16 bits, using the extra bits to encode the discriminator for the pointer V offset - and if it's just assertions that it's in the right mode, it won't make the static case any less efficient for this generality - and remove some complexity (eg: the DWARF5 name emission can be a non-template then))

If the DWARF5AccelTable*Data types become one data type, then I suppose this "finalize" operation could be a non-member function that takes a `AccelTable<DWARF5AccelTableData>` parameter which would be pretty simple/tidy.

Reply:
Thanks for suggestions. I expanded this refactor.

I removed DWARF5AccelTableStaticData (24 bytes), and changed to use DWARF5AccelTableData (32 bytes).
I added class DWARF5AccelTable : public AccelTable<DWARF5AccelTableData>
  -- to move getEntries into it, to reduce the scope of where it's called, and to reduce number of template instantiations.
Made the finalize function static that takes in DWARF5AccelTable

My reasoning for living std::variant in place for now:
  -- Data in above comments shows when we are compiling sema from bit code it is noise. 
  -- The DWARF5AccelTableStaticData is used in DWARF*Linker It raises entire debug information into an IR form, so extra 8 bytes for a acceleration table I don't think matter that much. For example in BOLT which uses similar approach peak memory is 50-60 GB with breaking up how many CUs we process in IR form before writing them out. Without it it's 100+GB. 
  -- Using variant vs bit packing is more inline with c++ usage in llvm (I think).
  
 WDYT?

Reply:
I'd probably still pack the structure a bit tighter - it's not especially costly for the abstractions, etc. But *shrug* I don't insist on it.